[CWS] Adding new bind event

spikat opened the pull request:
New bind event.

For now it only support `AF_INET` and `AF_INET6` address families, and it exposes:
- the used socket as integer: `bind.socket`
- the address family as string (using unix constants): `bind.addr.family`
- ip as string: `bind.addr`
- and port as integer: `bind.addr_port`

New SECL rules can be written like: `bind.port != 1234 && bind.addr.family == AF_INET && bind.addr != "127.0.0.1"`

We do not filter bind unsupported address families (in a sense where we do not export related sockaddr fields), but we still send them (with `addr_port == 0` and `addr == ""`).

Also, for now all calls to binds that fails will be send, no matter the related error code (only filtering `EACCESS` looks pity for bind)

Functional tests were added:
`inv -e security-agent.functional-tests --testflags "-test.v -test.run TestBindEvent"`

NB: It was decided that listen will not be added for now (cause we didn't see any use-case where it could be useful)

TODO:
- Adding more address family coverage (and find a nice way to do it without having all the possible fields on the same event json schema, with different empty fields on each event)
- Define which errors we want to discard


<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Draft PRs should be prefixed with `[WIP]` in their title.

-->
### What does this PR do?

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succint form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

spikat committed:
Adding new bind event.
SECL to be rework, for instance it exposes the socket, address family, ip and port for ipv4 and ipv6

TODO:
- Add functional tests
- Add more address family coverage
- Define which errors we want to discard

spikat committed:
Adding bind options to syscall_tester for AF_INET/AF_INET6

spikat committed:
Add functionnal tests

spikat committed:
Do not discard binds with unhandled address family

spikat committed:
Add functionnal test to validate unhandled address family event

spikat commented on pkg/security/secl/model/model.go:
```diff
@@ -796,6 +797,17 @@ type DNSEvent struct {
+	Socket     int32  `field:"socket"`      // the socket to be bound
+	AddrFamily uint16 `field:"addr_family"` // address family
+	AddrPort   uint16 `field:"port"`        // port (if any)
+	Addr       string `field:"addr"`        // address (if any)
```
Totally agree, will make the change

spikat replied:
I applied your suggestion! Thanks.

spikat committed:
Add ChangeLog

kayayarai commented on docs/cloud-workload-security/backend.schema.json:
```diff
@@ -137,6 +141,34 @@
+      "properties": {
+        "socket": {
+          "type": "integer",
+          "description": "bound socket"
```
```suggestion
          "description": "Bound socket"
```

kayayarai commented on docs/cloud-workload-security/backend.schema.json:
```diff
@@ -137,6 +141,34 @@
+        },
+        "addr_family": {
+          "type": "string",
+          "description": "address family"
```
```suggestion
          "description": "Address family"
```

spikat replied:
Applied, thanks!

kayayarai commented on docs/cloud-workload-security/backend.schema.json:
```diff
@@ -137,6 +141,34 @@
+        },
+        "addr_port": {
+          "type": "integer",
+          "description": "bound port (if any)"
```
```suggestion
          "description": "Bound port (if any)"
```

spikat replied:
Applied, thanks!

kayayarai commented on docs/cloud-workload-security/backend.schema.json:
```diff
@@ -137,6 +141,34 @@
+        },
+        "addr": {
+          "type": "string",
+          "description": "bound address (if any)"
```
```suggestion
          "description": "Bound address (if any)"
```

spikat replied:
Applied, thanks!

kayayarai commented on docs/cloud-workload-security/secl.json:
```diff
@@ -514,6 +514,40 @@
+        {
+          "name": "bind.addr",
+          "type": "string",
+          "definition": "address (if any)"
```
```suggestion
          "definition": "Address (if any)"
```

spikat replied:
Applied, thanks!

kayayarai commented on docs/cloud-workload-security/secl.json:
```diff
@@ -514,6 +514,40 @@
+        {
+          "name": "bind.addr_family",
+          "type": "int",
+          "definition": "address family"
```
```suggestion
          "definition": "Address family"
```

spikat replied:
Applied, thanks!

kayayarai commented on docs/cloud-workload-security/secl.json:
```diff
@@ -514,6 +514,40 @@
+        {
+          "name": "bind.port",
+          "type": "int",
+          "definition": "port (if any)"
```
```suggestion
          "definition": "Port (if any)"
```

spikat replied:
Applied, thanks!

kayayarai commented on docs/cloud-workload-security/secl.json:
```diff
@@ -514,6 +514,40 @@
+        {
+          "name": "bind.socket",
+          "type": "int",
+          "definition": "the socket to be bound"
```
```suggestion
          "definition": "The socket to be bound"
```

spikat replied:
Applied, thanks!

kayayarai commented on releasenotes/notes/runtime-security-bind-a885daad9b1e891c.yaml:
```diff
@@ -0,0 +1,18 @@
+#
+# Each section note must be formatted as reStructuredText.
+---
+upgrade:
+  - |
+    List upgrade notes here, or remove this section.
+    Upgrade notes should be rare: only list known/potential breaking changes,
+    or major behaviorial changes that require user action before the upgrade.
+    Notes here must include steps that users can follow to 1. know if they're
+    affected and 2. handle the change gracefully on their end.
```
```suggestion

```

spikat replied:
Applied, thanks!

kayayarai reviewed (changes requested):
Not sure if the .md also needs to have these capitalization edits made, but I suspect it's generated from the .json. Throughout, capitalize the first word of the descriptions.

spikat committed:
Review: fix documentation (mostly typo).

Co-authored-by: Kari Halsted <12926135+kayayarai@users.noreply.github.com>

spikat commented:
> 
HI @kayayarai, yes those files are generated and will be fixed as well grace to your other suggestions. :)

spikat committed:
Do not expose socket anymore

spikat committed:
Make use of the IPPortContext type for bind IP/Port fields

spikat committed:
Update documentation

spikat committed:
Tag the event as experimental (until we agree on a JSON output)

lebauce commented on pkg/security/probe/serializers.go:
```diff
@@ -317,6 +317,13 @@ type SpliceEventSerializer struct {
+// BindEventSerializer serializes a bind event to JSON
+// easyjson:json
+type BindEventSerializer struct {
+	AddrFamily string            `json:"addr_family" jsonschema_description:"Address family"`
```
Should we put "family" below the "addr" field like "ip" and "port" ?

spikat replied:
You're right, it's cleanest. Updated on last commit. Thanks

lebauce commented on pkg/security/ebpf/c/bind.h:
```diff
@@ -0,0 +1,105 @@
+        return 0;
+    }
+
+    int retval = PT_REGS_RC(ctx); /* TODO: define which errors we want to discard */
```
We should probably use IS_UNHANDLED_ERROR for now, as for the other events

spikat replied:
Applied on last commit. Thanks

lebauce commented on pkg/security/secl/model/consts_other.go:
```diff
@@ -21,4 +21,5 @@ var (
 	mmapFlagConstants         = map[string]int{}
 	mmapFlagArchConstants     = map[string]int{}
 	signalConstants           = map[string]int{}
+	addressFamilyuConstants   = map[string]int{}
```
s/addressFamilyuConstants/addressFamilyConstants/

spikat committed:
Review: apply suggestions:
- put the family under the "addr" sub-part
- filter unhandled errors (and remove corresponding tests)
- also fix a typo

lebauce commented on pkg/security/secl/model/model.go:
```diff
@@ -796,6 +797,16 @@ type DNSEvent struct {
+	SyscallEvent
+
+	Socket     int32         `field:"socket"`      // The socket to be bound
+	AddrFamily uint16        `field:"addr_family"` // Address family
```
s/addr_family/addr.familty/

spikat replied:
Updated on last commit. Thanks

spikat committed:
Review: update also the secl part for addr.family

Gui774ume commented on pkg/security/ebpf/c/bind.h:
```diff
@@ -0,0 +1,109 @@
+        .bind = {
+            .socket = socket,
+            .addr = addr,
+            .addr_len = addr_len,
```
I believe this field is unused.

spikat replied:
Updated, thanks.

Gui774ume commented on pkg/security/secl/model/model.go:
```diff
@@ -796,6 +797,16 @@ type DNSEvent struct {
+type BindEvent struct {
+	SyscallEvent
+
+	Socket     int32         `field:"socket"`      // The socket to be bound
```
Unless you thought of a use case, we should probably remove it for now: we don't export any file descriptor from any other event so I'm not sure what you can do with it.

spikat replied:
I'm not sure either. OK let's simplify it for now.

spikat replied:
Updated, thanks.

Gui774ume commented on pkg/security/ebpf/c/bind.h:
```diff
@@ -0,0 +1,109 @@
+    return 0;
+}
+
+SYSCALL_KRETPROBE(bind) {
```
You need to add the tracepoint hook points. These hook points are required to handle kernels that do not support the maxactive parameter. Checkout `tracepoint/syscalls/sys_exit_bpf` for an example with the bpf event.

spikat replied:
Oh right, thank you! Updated on last commit

Gui774ume commented on pkg/security/ebpf/c/bind.h:
```diff
@@ -0,0 +1,109 @@
+    /* get additionnal information, depending on the addr family */
+    struct sockaddr_in* sockin;
+    struct sockaddr_in6* sockin6;
+    switch (addr_family) {
```
This entire section seems to be a duplicate of `kprobe/security_socket_bind` in `flow.h`. You might want to keep only one of them / refactor the kprobe hook point if you want to make the code cleaner :)
This also means that you probably don't need to put anything in the syscall_cache, but I would still keep the syscall hook points so that we know that this bind event comes from a syscall.

Gui774ume commented on pkg/security/secl/model/category.go:
```diff
@@ -37,7 +37,7 @@ func GetEventTypeCategory(eventType eval.EventType) EventCategory {
 	case "exec", "signal", "exit", "fork":
 		return ProcessCategory
-	case "bpf", "selinux", "mmap", "mprotect", "ptrace", "load_module", "unload_module":
+	case "bpf", "selinux", "mmap", "mprotect", "ptrace", "load_module", "unload_module", "bind":
```
kernel or network ?

spikat replied:
Initially I was wondering and start with network, but it automatically adds the network context (src/dst addresses etc) which make no sense for bind; so I switch it back to kernel (but I'm still not sure, as the difference between kernel and process categories aren't clear for me :))

Gui774ume reviewed (commented):
Looks great ! I left a few comments, mainly in the eBPF part :)

spikat committed:
Review: remove unused fields (socket and addr_len)

spikat commented on pkg/security/secl/model/unmarshallers.go:
```diff
@@ -900,3 +901,31 @@
+		return 0, ErrNotEnoughData
+	}
+
+	e.Socket = int32(ByteOrder.Uint32(data[read : read+4]))
```
Indeed! Thanks.

spikat commented on pkg/security/tests/bind_test.go:
```diff
@@ -0,0 +1,123 @@
+		})
+	})
+
+	test.Run(t, "bind-af-unknown-unix", func(t *testing.T, kind wrapperType, cmdFunc func(cmd string, args []string, envs []string) *exec.Cmd) {
```
Not entirely, it shouldn't be a big deal retrieving it. But it raises some other questions:
- how did we handle the various other types of address?
- the unix socket is more like a file, how did we want to share events related to this socket in the future?

spikat committed:
Review: add the missing exit tracepoint hook point

spikat committed:
Review: factorise the addr parsing

spikat committed:
Review: try to not parse sockaddr twice

Commit:
enable security_socket_bind for bind events

Commit:
rebase main

Commit:
fix windows build
